store/src/java/com/zimbra/cs/service/account/EndSession.java

```diff
@@ -40,23 +47,56 @@
     public Element handle(Element request, Map<String, Object> context)
     throws ServiceException {
         ZimbraSoapContext zsc = getZimbraSoapContext(context);
-        if (zsc.hasSession()) {
-            Session s = getSession(zsc);
-            endSession(s);
-        }
-        boolean clearCookies = request.getAttributeBool(AccountConstants.A_LOG_OFF, false);
-        if (clearCookies || getAuthenticatedAccount(zsc).isForceClearCookies()) {
-            context.put(SoapServlet.INVALIDATE_COOKIES, true);
-            try {
-                AuthToken at = zsc.getAuthToken();
-                HttpServletRequest httpReq = (HttpServletRequest) context.get(SoapServlet.SERVLET_REQUEST);
-                HttpServletResponse httpResp = (HttpServletResponse) context.get(SoapServlet.SERVLET_RESPONSE);
-                at.encode(httpReq, httpResp, true);
-                at.deRegister();
-            } catch (AuthTokenException e) {
-                throw ServiceException.FAILURE("Failed to de-register an auth token", e);
+        boolean clearAllSessions = request.getAttributeBool(AccountConstants.A_CLEAR_ALL_SOAP_SESSIONS, false);
+        boolean excludeCurrrentSession = request.getAttributeBool(AccountConstants.A_EXCLUDE_CURRENT_SESSION, false);
+
+        Account account = getAuthenticatedAccount(zsc);
+        if (clearAllSessions) {
+            String currentSessionId = null;
+            if (excludeCurrrentSession && zsc.hasSession()) {
+                Session currentSession = getSession(zsc);
+                currentSessionId = currentSession.getSessionId();
+            }
+            Collection<Session> sessionCollection = SessionCache.getSoapSessions(account.getId());
+            if (sessionCollection != null) {
+                List<Session> sessions = new ArrayList<Session>(sessionCollection);
```

Comment:
Why not use for (Session session: sessions)

Reply:
was getting ConcurrentModificationException while using for (Session session: sessions)